openage/convert/dataformat/format_metadata.py

```diff
@@ -0,0 +1,154 @@
+        for layer_id, kwargs in self.data['layers'].items():
+            self.add_layer(layer_id, **kwargs)
+        for degree, args in self.data['angles'].items():
+            self.add_angle(degree, args['mirror_from'])
+            for frame in args['frames']:
+                self.add_frame(*frame)
+
+
+class TerrainDefinition(DataDefinition):
+    type = 'terrain'
+    version = 0
+
+    def add_blending_priority(self, priority):
+        self.txt.append('blending_priority {}\n'.format(priority))
+
+    def add_blending_mask(self, blend_id, filename):
+        self.txt.append('blending_mask {} {}\n'.format(blend_id, filename))
+
+    def add_dots_per_tile(self, dots_number):
+        self.txt.append('dots_per_tile {}\n'.format(dots_number))
+
+    def format_data(self):
+        self.add_header()
+        self.add_version()
+        self.add_blending_priority(self.data['blending_priority'])
+        self.add_blending_mask(**self.data['blending_mask'])
+        self.add_dots_per_tile(self.data['dots_per_tile'])
+        for image_id, file in self.data['image_files'].items():
+            self.add_image(image_id, file)
+        for layer_id, kwargs in self.data['layers'].items():
+            self.add_layer(layer_id, **kwargs)
+        for frame in self.data['frames']:
+            self.add_frame(*frame)
+
+
+if __name__ == '__main__':
+    dummy_sprite_data = {
+        'image_files': {
+            '0': 'image_file0',
+            '1': 'image_file1'
+        },
+        'layers': {
+            '0': {'mode': 'off', 'position': 'shadow'},
+            '1': {'mode': 'once', 'position': 'roof', 'time_per_frame': '1.0'},
+            '2': {'mode': 'loop', 'position': 'default', 'time_per_frame': '1.0', 'replay_delay': '2.0'}
+        },
+        'angles': {
+            '90': {'mirror_from': None, 'frames': [
+                ('0', '0', '3', '5', '40', '60', '20', '20'),
+                ('0', '0', '33', '35', '70', '90', '50', '50'),
+            ]},
+            '180': {'mirror_from': '90', 'frames': []},
+            '270': {'mirror_from': None, 'frames': [
+                ('3', '2', '5', '6', '23', '90', '33', '12'),
+                ('2', '1', '22', '12', '34', '55', '97', '10'),
+            ]}
+        }
+    }
+
+    dummy_terrain_data = {
+        'blending_priority': '2',
+        'blending_mask': {'blend_id': '0', 'filename': 'blend_file'},
+        'dots_per_tile': '2.0',
+        'image_files': {
+            '0': 'image_file0',
+            '1': 'image_file1'
+        },
+        'layers': {
+            '0': {'mode': 'off'},
+            '1': {'mode': 'once', 'time_per_frame': '1.0'},
+            '2': {'mode': 'loop', 'time_per_frame': '1.0', 'replay_delay': '2.0'}
+        },
+        'frames': [
+                ('0', '3', '5', '40', '60'),
+                ('0', '33', '35', '70', '90'),
+                ('3', '5', '6', '23', '90'),
+                ('2', '22', '12', '34', '55'),
+            ]
```

Comment:
I'm actually not sure if I'm assuming a wrong data format here. Does terrain frame information have NO angle data, or does it just all have the same angle? The same question goes for the other "missing" data, xhotspot and yhotspot. By looking and the Sprite definition it's unclear.

Reply:
In general, should I expect the input data to have a standardized shape, or should I take care of the exceptions here?

And if it is standardized, where should I expect `sprite` and `terrain`'s data to have the same format, and where not?

Reply:
Yes, there are no angles for terrains. They don't need them. Units can rotate and should show different sprites depending on their relative angle to the camera position. Terrains are just ground textures. They also do not need {x|y}_hotspot because they are not attached to anything.

> In general, should I expect the input data to have a standardized shape, or should I take care of the exceptions here?

You should force a shape via the method's requested parameters. Then you can transfer them in whatever storage format fits the most.

> And if it is standardized, where should I expect sprite and terrain's data to have the same format, and where not?

I think the best approach is to think of `terrain` and `sprite` as different formats that just happen to have some similarities *now*. We will probably change them quite a bit during development and they could still diverge depending on what we need.